flaky test TestStaleReadAllCombinations

```
=== NAME  TestStaleReadAllCombinations
    result.go:49: 
        	Error Trace:	pkg/testkit/result.go:49
        	            				tests/realtikvtest/txntest/stale_read_test.go:1717
        	Error:      	Not equal: 
        	            	expected: "[1 10]\n"
        	            	actual  : "[1 10]\n[2 20]\n"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,3 @@
        	            	 [1 10]
        	            	+[2 20]
        	            	 
        	Test:       	TestStaleReadAllCombinations
        	Messages:   	sql:select * from t, args:[]
=== NAME  TestStaleReadAllCombinations/with_multiple_labels/closest-adaptive/tidb_read_staleness
    testing.go:1577: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
--- FAIL: TestStaleReadAllCombinations/with_multiple_labels/closest-adaptive/tidb_read_staleness (0.00s)
```

https://do.pingcap.net/jenkins/blue/rest/organizations/jenkins/pipelines/pingcap/pipelines/tidb/pipelines/ghpr_check2/runs/39527/nodes/106/steps/662/log/?start=0

--- a/tests/realtikvtest/txntest/stale_read_test.go
+++ b/tests/realtikvtest/txntest/stale_read_test.go
@@ -1493,6 +1493,7 @@ func TestStaleReadAllCombinations(t *testing.T) {
 	defer config.RestoreFunc()()
 
 	tk := testkit.NewTestKit(t, store)
+	oracleClient := tk.Session().GetStore().GetOracle()
 
 	safePointName := "tikv_gc_safe_point"
 	safePointValue := "20160102-15:04:05 -0700"
@@ -1508,20 +1509,24 @@ func TestStaleReadAllCombinations(t *testing.T) {
 
 	// Insert row #1
 	tk.MustExec("insert into t values (1, 10)")
-	time.Sleep(3000 * time.Millisecond)                   // Increased from 1s to 3s for better timing separation
-	firstTime := time.Now().Add(-1500 * time.Millisecond) // Adjusted offset
-	// Retrieve current TSO from store's Oracle instead of @@tidb_current_ts.
-	externalTS, err := store.GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
-	if err != nil {
-		t.Fatalf("failed to get TSO: %v", err)
-	}
+	time.Sleep(3000 * time.Millisecond) // Increased from 1s to 3s for better timing separation
+	ts, err := oracleClient.GetTimestamp(context.Background(), &oracle.Option{})
+	require.NoError(t, err)
+	firstTime := oracle.GetTimeFromTS(ts).Add(-1500 * time.Millisecond)
+
+	externalTS, err := oracleClient.GetTimestamp(context.Background(), &oracle.Option{})
+	require.NoError(t, err)
 
 	time.Sleep(3000 * time.Millisecond) // Increased from 1s to 3s
 	// Insert row #2
 	tk.MustExec("insert into t values (2, 20)")
-	row2CreatedTime := time.Now()
-	time.Sleep(3000 * time.Millisecond)                    // Increased from 1s to 3s
-	secondTime := time.Now().Add(-1500 * time.Millisecond) // Adjusted offset
+	row2CreatedTS, err := oracleClient.GetTimestamp(context.Background(), &oracle.Option{})
+	require.NoError(t, err)
+	row2CreatedTime := oracle.GetTimeFromTS(row2CreatedTS)
+	time.Sleep(3000 * time.Millisecond) // Increased from 1s to 3s
+	ts, err = oracleClient.GetTimestamp(context.Background(), &oracle.Option{})
+	require.NoError(t, err)
+	secondTime := oracle.GetTimeFromTS(ts).Add(-1500 * time.Millisecond)
 
 	staleReadMethods := []struct {
 		name   string
@@ -1533,7 +1538,10 @@ func TestStaleReadAllCombinations(t *testing.T) {
 		{
 			name: "tidb_read_staleness",
 			setup: func() {
-				row2CreatedElapsed := int(time.Since(row2CreatedTime).Seconds())
+				ts, err := oracleClient.GetTimestamp(context.Background(), &oracle.Option{})
+				require.NoError(t, err)
+				elapsed := oracle.GetTimeFromTS(ts).Sub(row2CreatedTime)
+				row2CreatedElapsed := int(elapsed.Seconds())
 				// Ensure sufficient buffer: with 3s sleep intervals, elapsed should be ~3s
 				// Adding 2s buffer ensures stale read is safely before row2 insertion
 				staleness := row2CreatedElapsed + 2

--- a/tests/realtikvtest/txntest/stale_read_test.go
+++ b/tests/realtikvtest/txntest/stale_read_test.go
@@ -1508,20 +1508,20 @@ func TestStaleReadAllCombinations(t *testing.T) {
 
 	// Insert row #1
 	tk.MustExec("insert into t values (1, 10)")
-	time.Sleep(1000 * time.Millisecond)
-	firstTime := time.Now().Add(-500 * time.Millisecond)
+	time.Sleep(3000 * time.Millisecond)                   // Increased from 1s to 3s for better timing separation
+	firstTime := time.Now().Add(-1500 * time.Millisecond) // Adjusted offset
 	// Retrieve current TSO from store's Oracle instead of @@tidb_current_ts.
 	externalTS, err := store.GetOracle().GetTimestamp(context.Background(), &oracle.Option{})
 	if err != nil {
 		t.Fatalf("failed to get TSO: %v", err)
 	}
 
-	time.Sleep(1000 * time.Millisecond)
+	time.Sleep(3000 * time.Millisecond) // Increased from 1s to 3s
 	// Insert row #2
 	tk.MustExec("insert into t values (2, 20)")
 	row2CreatedTime := time.Now()
-	time.Sleep(1000 * time.Millisecond)
-	secondTime := time.Now().Add(-500 * time.Millisecond)
+	time.Sleep(3000 * time.Millisecond)                    // Increased from 1s to 3s
+	secondTime := time.Now().Add(-1500 * time.Millisecond) // Adjusted offset
 
 	staleReadMethods := []struct {
 		name   string
@@ -1534,7 +1534,9 @@ func TestStaleReadAllCombinations(t *testing.T) {
 			name: "tidb_read_staleness",
 			setup: func() {
 				row2CreatedElapsed := int(time.Since(row2CreatedTime).Seconds())
-				staleness := row2CreatedElapsed + 1 // The time `now - staleness(second)` is between row1 and row2.
+				// Ensure sufficient buffer: with 3s sleep intervals, elapsed should be ~3s
+				// Adding 2s buffer ensures stale read is safely before row2 insertion
+				staleness := row2CreatedElapsed + 2
 				tk.MustExec(fmt.Sprintf("set @@tidb_read_staleness='-%d'", staleness))
 			},
 			query: "select * from t",

--- a/tests/realtikvtest/txntest/stale_read_test.go
+++ b/tests/realtikvtest/txntest/stale_read_test.go
@@ -1643,6 +1643,9 @@ func TestStaleReadAllCombinations(t *testing.T) {
 		},
 	}
 
+	tidbReadStalenessMethod := staleReadMethods[0]
+	otherStaleReadMethods := staleReadMethods[1:]
+
 	replicaReadSettings := []string{
 		"leader",
 		"follower",
@@ -1695,30 +1698,45 @@ func TestStaleReadAllCombinations(t *testing.T) {
 		*/
 	}
 
-	// Test all combinations
+	// Run tidb_read_staleness combinations first to minimize elapsed time from setup.
 	for _, label := range labelSettings {
-		t.Run(label.name, func(t *testing.T) {
-			// Update global config with current label setting
+		t.Run(label.name+"/tidb_read_staleness", func(t *testing.T) {
 			conf := *config.GetGlobalConfig()
 			conf.Labels = label.labels
 			config.StoreGlobalConfig(&conf)
 
 			for _, replicaRead := range replicaReadSettings {
 				t.Run(replicaRead, func(t *testing.T) {
-					// Set replica read mode
+					tk.MustExec(fmt.Sprintf("set @@tidb_replica_read='%s'", replicaRead))
+					method := tidbReadStalenessMethod
+					t.Run(method.name, func(t *testing.T) {
+						defer method.clean()
+						method.setup()
+						result := tk.MustQuery(method.query)
+						result.Check(testkit.Rows(method.expect...))
+					})
+				})
+			}
+		})
+	}
+
+	// Run remaining stale read methods and transaction mode combinations afterwards.
+	for _, label := range labelSettings {
+		t.Run(label.name+"/others", func(t *testing.T) {
+			conf := *config.GetGlobalConfig()
+			conf.Labels = label.labels
+			config.StoreGlobalConfig(&conf)
+
+			for _, replicaRead := range replicaReadSettings {
+				t.Run(replicaRead, func(t *testing.T) {
 					tk.MustExec(fmt.Sprintf("set @@tidb_replica_read='%s'", replicaRead))
 
-					for _, method := range staleReadMethods {
+					for _, method := range otherStaleReadMethods {
 						t.Run(method.name, func(t *testing.T) {
-							// Setup stale read method
 							defer method.clean()
 							method.setup()
-
-							// Execute query and verify results
 							result := tk.MustQuery(method.query)
 							result.Check(testkit.Rows(method.expect...))
-
-							// Cleanup
 						})
 					}
 

